keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ArtifactTypeConstraintEvaluator.kt

```diff
@@ -0,0 +1,33 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.core.api.ArtifactTypeConstraint
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+@Component
+class ArtifactTypeConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher
+) : ConstraintEvaluator<ArtifactTypeConstraint> {
+  override val supportedType = SupportedConstraintType<ArtifactTypeConstraint>("artifact-type")
+  override fun isImplicit() = true
+
+  override fun canPromote(artifact: DeliveryArtifact, version: String, deliveryConfig: DeliveryConfig, targetEnvironment: Environment): Boolean {
+    val allowedTypes = mutableSetOf<ArtifactType>()
+    deliveryConfig
+      .environments
+      .firstOrNull { it.name == targetEnvironment.name }
+      ?.resources
+      ?.forEach { resource ->
+        val usedArtifact = resource.findAssociatedArtifact(deliveryConfig)
+        if (usedArtifact != null) {
+          allowedTypes.add(usedArtifact.type)
```

Comment:
If you do the same sort of thing to build an `allowedArtifacts` list, and asserted that `artifact` is in the list, this would also solve #833.

Reply:
@asher GREAT THOUGHT!